packages/react/src/components/Tag/index.tsx

```diff
@@ -18,14 +23,45 @@ TagLabel.propTypes = {
   className: PropTypes.string
 };
 
-const Tag = ({ children, className, ...other }: TagProps) => (
-  <div className={classNames('Tag', className)} {...other}>
-    {children}
-  </div>
-);
+const Tag = ({
+  children,
+  className,
+  variant,
+  onDismiss,
+  ...other
+}: TagProps) => {
+  const [show, setShow] = useState(true);
+  const Component = variant === 'dismiss' ? 'button' : 'div';
+
+  const handleDismiss = (event: React.MouseEvent<HTMLButtonElement>) => {
+    setShow(false);
+    if (onDismiss) {
+      onDismiss(event);
+    }
+  };
+
+  if (variant === 'dismiss') {
+    (other as any).onClick = handleDismiss;
+  }
+
+  return (
+    <Component
+      className={classNames('Tag', className, {
+        'Tag--hidden': !show,
+        'Tag--dismiss': variant === 'dismiss'
+      })}
+      {...other}
+    >
+      {children}
+      {variant === 'dismiss' && <Icon type="close" />}
+    </Component>
+  );
```

Comment:
I think this could be simplified to something like this:

```suggestion
  return (
    <Component
      className={classNames('Tag', className, {
        'Tag--hidden': !show,
        'Tag--dismiss': variant === 'dismiss'
      })}
      {...other}
      onClick={handleDismiss}
    >
      {children}
      {variant === 'dismiss' && <Icon type="close" />}
    </Component>
  );
```

Reply:
Combined the two variant prs into one and used Union type for better typing!